storage/database/sharded_database.go

```diff
@@ -151,102 +154,283 @@ func (pdb *shardedDB) Close() {
+type shardedDBIteratorUnsorted struct {
+	shardedDBIterator
+}
+
+// NewIteratorUnsorted creates a iterator over the entire keyspace contained within
+// the key-value database.
+// If you want to get ordered items in serial, checkout shardedDB.NewIterator()
+// If you want to get items in parallel from channels, checkout shardedDB.NewChanIterator()
+func (db *shardedDB) NewIteratorUnsorted() Iterator {
+	return db.newIteratorUnsorted(func(db Database) Iterator { return db.NewIterator() })
+}
+
+// NewIteratorWithStartUnsorted creates a iterator over a subset of database content
+// starting at a particular initial key (or after, if it does not exist).
+func (db *shardedDB) NewIteratorWithStartUnsorted(start []byte) Iterator {
+	return db.newIteratorUnsorted(func(db Database) Iterator { return db.NewIteratorWithStart(start) })
+}
+
+// NewIteratorWithPrefixUnsorted creates a iterator over a subset of database content
+// with a particular key prefix.
+func (db *shardedDB) NewIteratorWithPrefixUnsorted(prefix []byte) Iterator {
+	return db.newIteratorUnsorted(func(db Database) Iterator { return db.NewIteratorWithPrefix(prefix) })
+}
+
+func (db *shardedDB) newIteratorUnsorted(newIterator func(Database) Iterator) Iterator {
+	it := &shardedDBIteratorUnsorted{
+		shardedDBIterator{
+			db.NewChanIterator(context.Background(), newIterator),
+			make(chan entry, shardedDBCombineChanSize),
+			nil, nil}}
+
+	go it.newCombineWorker()
+
+	return it
+}
+
+// newCombineWorker fetches any key/value from resultChs and put the data in resultCh
+func (it shardedDBIteratorUnsorted) newCombineWorker() {
+Iter:
+	for len(it.resultChs) != 0 {
+	chanIter:
+		for i, ch := range it.resultChs {
+			select {
+			case <-it.ctx.Done():
+				logger.Trace("[shardedDBIteratorUnsorted] combine worker ends due to ctx")
+				break Iter
+			case e, ok := <-ch:
+				if !ok { // channel is closed
+					it.resultChs = append(it.resultChs[:i], it.resultChs[i+1:]...)
+					break chanIter
+				} else {
+					it.resultCh <- e
+				}
+			default:
+				// no item in ch
+			}
+		}
+	}
+	logger.Trace("[shardedDBIteratorUnsorted] combine worker finishes")
+	close(it.resultCh)
+}
+
+// shardedDBChanIterator creates iterators for each shard DB.
+// Channels subscribing each iterators can be gained.
+// Each iterators fetch values in binary-alphabetical order.
+// This is useful when you want to operate on each items in parallel.
+type shardedDBChanIterator struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	iterators []Iterator
+	resultChs []chan entry
+}
+
+// NewChanIterator creates iterators for each shard DB.
+// This is useful when you want to operate on each items in parallel.
+// If you want to get ordered items in serial, checkout shardedDB.NewIterator()
+// If you want to get unordered items in serial, checkout shardedDB.NewChanIteratorUnsorted().
+func (db *shardedDB) NewChanIterator(ctx context.Context, newIterator func(Database) Iterator) shardedDBChanIterator {
```

Comment:
`NewChanIterator` seems to be used only in the database package, as it is wrapped by `newIterator`, is there any reason for making it public?

Reply:
Also, `Channels` and `Release` should be wrapped by the function of `shardedDBIterator` or `shardedDBIteratorUnSorted`, they should be private, shouldn't they?

Reply:
I thought `shardedDBChanIterator` is needed more than `shardedDBIterator ` or `shardedDBIteratorUnSorted `.
If we are using `shardedDBIterator ` or `shardedDBIteratorUnSorted `, all data is carried with only one channel. If we use `shardedDBChanIterator`, data is carried with several channels and some operations could be done in parallel. I am considering using `shardedDBChanIterator` in DB migration for performance. I expect there could be other cases to use `shardedDBChanIterator` than `shardedDBIterator ` or `shardedDBIteratorUnSorted `.

Reply:
I understood. Thanks!